File: BackEnd/Utils.Notifications/NotificationsPubNub.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using PubNubMessaging.Core;

namespace Utils.Notifications
{
    public class NotificationsPubNub : INotifications
    {
        private static string PublishKey = "pub-c-a024c6ed-96da-40cd-8e1a-45279bd4b63b";
        private static string SuscribeKey = "sub-c-7d3f5b56-6e54-11e6-9259-0619f8945a4f";

        private static Pubnub _pubnubCliente;

        public NotificationsPubNub()
        {
            _pubnubCliente = new Pubnub(PublishKey, SuscribeKey);
        }



        /// <summary>
        /// Channel names are UTF-8 compatible. Prohibited chars in a channel name are:
        /// comma: ,
        /// slash: /
        /// backslash: \
        /// period: .
        /// asterisks: *
        /// colon: :
        /// </summary>
        /// <param name="channelName"></param>
        public void SubscribeChanel(string channelName)
        {
            _pubnubCliente.Subscribe<string>(channel: channelName, subscribeCallback: DisplaySubscribeReturnMessage, 
                connectCallback: DisplaySubscribeConnectStatusMessage, errorCallback: DisplayErrorMessage);
        }

        public void UnsubscribeChanel(string channelName)
        {
            _pubnubCliente.Unsubscribe<string>(channel: channelName, subscribeCallback: DisplayReturnMessage, connectCallback: DisplaySubscribeConnectStatusMessage
                 , disconnectCallback: DisplaySubscribeDisconnectStatusMessage, errorCallback: DisplayErrorMessage);
        }

        public void SendMessage(string channelName, string message)
        {
            _pubnubCliente.Publish<string>(channel: channelName, message: message, userCallback: DisplayReturnMessage, errorCallback: DisplayErrorMessage);
        }


        /// <summary>
        /// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        /// </summary>
        /// 


        //public NotificationsPubNub()
        //{
        //    _pubnubCliente = new Pubnub(PublishKey, SuscribeKey);
        //    _pubnubCliente.Subscribe<string>(
        //channel: "Channel-15t1x43ar",
        //subscribeCallback: DisplaySubscribeReturnMessage,
        //connectCallback: DisplaySubscribeConnectStatusMessage,
        //errorCallback: DisplayErrorMessage);
        //}

        public void DisplaySubscribeDisconnectStatusMessage(string var)
        {
           // Console.WriteLine("Estado de desconexion: " + var);
        }
        public void DisplaySubscribeConnectStatusMessage(string connectMessage)
        {
           //Console.WriteLine("Estado de suscripcion: "+connectMessage);
            //_pubnubCliente.Publish<string>(
            //    channel: "Channel-15t1x43ar",
            //    message: "Soy Andres",
            //    userCallback: DisplayReturnMessage,
            //    errorCallback: DisplayErrorMessage);
        }

        void DisplaySubscribeReturnMessage(string result)
        {
            //Console.WriteLine("SUBSCRIBE REGULAR CALLBACK:");
            //Console.WriteLine("Resultado de suscribirse: "+result);
            if (!string.IsNullOrEmpty(result) && !string.IsNullOrEmpty(result.Trim()))
            {
                List<object> deserializedMessage = _pubnubCliente.JsonPluggableLibrary.DeserializeToListOfObject(result);
                if (deserializedMessage != null && deserializedMessage.Count > 0)
                {
                    object subscribedObject = (object)deserializedMessage[0];
                    if (subscribedObject != null)
                    {
                        //IF CUSTOM OBJECT IS EXCEPTED, YOU CAN CAST THIS OBJECT TO YOUR CUSTOM CLASS TYPE
                        string resultActualMessage = _pubnubCliente.JsonPluggableLibrary.SerializeToJsonString(subscribedObject);
                    }
                }
            }
        }

        void DisplayErrorMessage(PubnubClientError pubnubError)
        {
          //  Console.WriteLine(pubnubError.StatusCode);
        }

        void DisplayReturnMessage(string result)
        {
            //Console.WriteLine("PUBLISH STATUS CALLBACK");
           // Console.WriteLine(result);
        }


    }
}

File: FrontEnd/MemsysAndroid/app/src/main/java/pis16/memsysandroid/notifications/NotificationsPubNub.java
package pis16.memsysandroid.notifications;

import java.util.ArrayList;
import java.util.Hashtable;

import com.pubnub.api.*;
import org.json.*;
import pis16.memsysandroid.notifications.interfaces.INotifications;
import pis16.memsysandroid.ui.MainActivity;

/**
 * Created by marccio on 29-Aug-16.
 */
public class NotificationsPubNub implements INotifications {

    private static Pubnub pubnub;
    private String  PublishKey="pub-c-a024c6ed-96da-40cd-8e1a-45279bd4b63b";
    private String SuscribeKey="sub-c-7d3f5b56-6e54-11e6-9259-0619f8945a4f";
    MainActivity caller;
    public NotificationsPubNub(final MainActivity caller){
        this.caller = caller;
        pubnub = new Pubnub(PublishKey, SuscribeKey);

    }
    @Override
    public void SubscribeChannel(String channelName) {
        try {
            pubnub.subscribe(channelName, new Callback() {
                        @Override
                        public void connectCallback(String channelName, Object message) {
                            pubnub.publish(channelName, "Hello from the PubNub Java SDK", new Callback() {});
                        }

                        @Override
                        public void disconnectCallback(String channel, Object message) {
                            System.out.println("SUBSCRIBE : DISCONNECT on channel:" + channel
                                    + " : "
                                    + message.toString());
                        }

                        public void reconnectCallback(String channel, Object message) {
                            System.out.println("SUBSCRIBE : RECONNECT on channel:" + channel
                                    + " : "
                                    + message.toString());
                        }

                        @Override
                        public void successCallback(String channel, Object message) {
                            System.out.println("SUBSCRIBE : " + channel + " : "
                                    +  message.toString());
                            caller.onMessageReceived(message.toString());
                        }

                        @Override
                        public void errorCallback(String channel, PubnubError error) {
                            System.out.println("SUBSCRIBE : ERROR on channel " + channel
                                    + " : " + error.toString());
                        }
                    }
            );
        } catch (PubnubException e) {
            System.out.println(e.toString());
        }
    }

    @Override
    public  void SendMessage(String channelName, String message) {
        Callback callback = new Callback() {
            public void successCallback(String channel, Object response) {
                System.out.println(response.toString());
            }
            public void errorCallback(String channel, PubnubError error) {
                System.out.println(error.toString());
            }
        };
        pubnub.publish(channelName, message , callback);
    }

    @Override
    public void UnsubscribeChannel(String channelName) {
        pubnub.unsubscribe(channelName);
    }

    @Override
    public boolean isConnected() {
        Callback callback = new Callback() {
            public void successCallback(String channel, Object response) {
                System.out.println(response.toString());
            }
            public void errorCallback(String channel, PubnubError error) {
                System.out.println(error.toString());
            }
        };
        pubnub.time(callback);
        return true;
    }




}
